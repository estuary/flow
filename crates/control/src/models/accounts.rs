use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

use crate::models::Id;

/// Accounts are the owners of resources and the actors performing actions
/// within the system. An Account may represent a human, a company, or an
/// automated process.
///
/// An `Account` is a Rust representation of the Postgres database understanding
/// of the account.
#[derive(Clone, Debug, Deserialize, FromRow, Serialize)]
pub struct Account {
    /// When this record was created.
    pub created_at: DateTime<Utc>,
    /// A freeform representation of the name. eg. "Wiley Coyote" or "Acme Anvil Corp."
    pub display_name: String,
    /// An email address to contact the account owner.
    pub email: String,
    /// Primary key for this record.
    pub id: Id,
    /// The catalog name for this account. Eg. "wiley" or "acmeCo"
    pub name: String,
    /// A case-folded, unicode normalized version of the catalog name. This is
    /// used to ensure that accounts have _functionally_ unique names, while
    /// always deferring to `name` for display purposes.
    pub unique_name: String,
    /// When this record was last updated.
    pub updated_at: DateTime<Utc>,
}

/// NewAccount represents the data required to insert a new Account, with
/// remaining fields from `Account` generated by Postgres.
#[derive(Debug, Deserialize, FromRow, Serialize)]
pub struct NewAccount {
    /// A freeform representation of the name. eg. "Wiley Coyote" or "Acme Anvil Corp."
    pub display_name: String,
    /// An email address to contact the account owner.
    pub email: String,
    /// A raw representation of the catalog name for this account. Eg. "wiley" or "acmeCo"
    pub name: String,
}
