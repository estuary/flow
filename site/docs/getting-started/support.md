
# Support

There are multiple ways to receive support if you run into issues or have questions while working with Estuary Flow.

:::tip
Be aware of downtime affecting Estuary components by subscribing to [status updates](https://status.estuary.dev/).
:::

## How to reach us

* Use [Slack](https://go.estuary.dev/slack) for general questions about the product.

* [Email us](mailto:support@estuary.dev) if your question or issue includes information specific to your company.

* You can also [report issues](https://github.com/estuary/flow/issues) directly in our GitHub repo.

## Grant data sharing access for troubleshooting

During certain troubleshooting sessions, it can be helpful for a member of the Estuary team to have access to your tenant data. An Estuary team member may request an access grant to help you manage your pipelines at these times.

:::warning
Estuary will _only_ ever request access for the `estuary_support/` prefix.
:::

You can temporarily grant Estuary access using the following steps:

1. Go to the [Admin page](https://dashboard.estuary.dev/admin/) in your Estuary dashboard.

2. Under the **Account Access** tab, find the **Data Sharing** table.

3. Select **Grant Access** and fill out the following details in the **Share Data** modal:

    * **Shared Prefix:** your tenant prefix
    * **Shared With:** `estuary_support/`
    * **Capability:** may depend on the troubleshooting session, but typically `admin`

4. Click **Grant Access** to confirm.

You should then see `estuary_support/` in the list of entities that you're sharing data with.

When your issue is resolved, you can revoke access again by selecting the `estuary_support/` row and clicking **Remove**.

Learn more about [access grants](../concepts/web-app.md#account-access).
