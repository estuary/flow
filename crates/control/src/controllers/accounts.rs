use axum::extract::Extension;
use axum::extract::Path;
use axum::response::IntoResponse;
use axum::Json;
use hyper::StatusCode;

use crate::context::AppContext;
use crate::error::AppError;
use crate::middleware::sessions::CurrentAccount;
use crate::models::accounts::Account;
use crate::models::accounts::NewAccount;
use crate::models::Id;
use crate::repo::accounts as accounts_repo;

pub mod routes;
mod view;

pub async fn index(Extension(ctx): Extension<AppContext>) -> Result<impl IntoResponse, AppError> {
    let accounts: Vec<Account> = accounts_repo::fetch_all(ctx.db()).await?;

    Ok((StatusCode::OK, view::index(accounts)))
}

pub async fn create(
    Extension(ctx): Extension<AppContext>,
    Json(input): Json<NewAccount>,
) -> Result<impl IntoResponse, AppError> {
    let account = accounts_repo::insert(ctx.db(), input).await?;

    Ok((StatusCode::CREATED, view::show(account)))
}

pub async fn show(
    Extension(ctx): Extension<AppContext>,
    Path(account_id): Path<Id>,
    CurrentAccount(current_account): CurrentAccount,
) -> Result<impl IntoResponse, AppError> {
    if account_id == current_account.id {
        let account = accounts_repo::fetch_one(ctx.db(), account_id).await?;

        Ok((StatusCode::OK, view::show(account)))
    } else {
        Err(AppError::AccessDenied)
    }
}
