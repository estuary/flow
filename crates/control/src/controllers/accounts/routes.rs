use crate::controllers::url_for;
use crate::models::accounts::Account;
use crate::models::id::Id;

pub fn index() -> String {
    url_for("/accounts")
}

pub fn show(account_id: Id<Account>) -> String {
    url_for(format!("/accounts/{}", account_id.to_string()))
}
