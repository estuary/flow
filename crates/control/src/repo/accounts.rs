use futures::TryFutureExt;
use sqlx::PgPool;

use crate::models::accounts::{Account, NewAccount};
use crate::models::Id;

pub async fn fetch_all(db: &PgPool) -> Result<Vec<Account>, sqlx::Error> {
    sqlx::query_as!(
        Account,
        r#"
    SELECT id as "id!: Id", display_name, email, name, unique_name, created_at, updated_at
    FROM accounts
    "#
    )
    .fetch_all(db)
    .await
}

pub async fn fetch_one(db: &PgPool, id: Id) -> Result<Account, sqlx::Error> {
    sqlx::query_as!(
        Account,
        r#"
    SELECT id as "id!: Id", display_name, email, name, unique_name, created_at, updated_at
    FROM accounts
    WHERE id = $1
    "#,
        id as Id
    )
    .fetch_one(db)
    .await
}

pub async fn insert(db: &PgPool, input: NewAccount) -> Result<Account, sqlx::Error> {
    // TODO: normalize the name
    let unique_name = input.name.clone();

    sqlx::query!(
        r#"
    INSERT INTO accounts(display_name, email, name, unique_name, created_at, updated_at)
    VALUES ($1, $2, $3, $4, NOW(), NOW())
    RETURNING id as "id!: Id"
    "#,
        input.display_name,
        input.email.to_lowercase(),
        input.name,
        unique_name,
    )
    .fetch_one(db)
    .and_then(|row| fetch_one(db, row.id))
    .await
}
