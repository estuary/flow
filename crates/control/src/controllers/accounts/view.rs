use axum::Json;

use crate::controllers::accounts::routes;
use crate::controllers::{
    json_api::DocumentData, json_api::Links, json_api::Many, json_api::One, json_api::Resource,
};
use crate::models::accounts::Account;

pub fn index(accounts: Vec<Account>) -> Json<Many<Account>> {
    let resources = accounts.into_iter().map(Resource::from).collect();
    let links = Links::default().put("self", routes::index());

    Json(DocumentData::new(resources, links))
}

pub fn show(account: Account) -> Json<One<Account>> {
    let payload = DocumentData::new(Resource::<Account>::from(account), Links::default());

    Json(payload)
}

impl From<Account> for Resource<Account> {
    fn from(account: Account) -> Self {
        let links = Links::default().put("self", routes::show(account.id));

        Resource {
            id: account.id,
            r#type: "account",
            attributes: account,
            links,
        }
    }
}
