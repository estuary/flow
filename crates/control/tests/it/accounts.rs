use control::repo::accounts::fetch_all;

use crate::support::redactor::Redactor;
use crate::support::{self, factory, test_context};

#[tokio::test]
async fn index_test() {
    // Arrange
    let mut t = test_context!();
    let account = factory::BatmanAccount.create(t.db()).await;
    t.login(account.clone());

    // Act
    let mut response = t.get("/accounts").await;

    // Assert
    assert!(response.status().is_success());
    let redactor = Redactor::default().redact(account.id, "a1");
    assert_json_snapshot!(redactor.response_json(&mut response).await.unwrap(), {
        ".data.*.attributes.created_at" => "[datetime]",
        ".data.*.attributes.updated_at" => "[datetime]",
    });
}

#[tokio::test]
async fn create_test() {
    // Arrange
    let mut t = test_context!();
    let account = factory::JokerAccount.create(t.db()).await;
    t.login(account.clone());
    let input = factory::BatmanAccount.attrs();

    // Act
    let mut response = t.post("/accounts", &input).await;

    // Assert
    let accounts = fetch_all(t.db())
        .await
        .expect("to insert test data")
        .into_iter()
        .filter(|a| a.id != account.id)
        .collect::<Vec<_>>();
    assert_eq!(1, accounts.len());
    assert_eq!(201, response.status().as_u16());
    let redactor = Redactor::default().redact(accounts[0].id, "a1");
    assert_json_snapshot!(redactor.response_json(&mut response).await.unwrap(), {
        ".data.attributes.created_at" => "[datetime]",
        ".data.attributes.updated_at" => "[datetime]",
    });
}

#[tokio::test]
async fn show_test() {
    // Arrange
    let mut t = test_context!();
    let account = factory::BatmanAccount.create(t.db()).await;
    t.login(account.clone());

    // Act
    let mut response = t.get(&format!("/accounts/{}", account.id)).await;

    // Assert
    assert!(response.status().is_success());
    let redactor = Redactor::default().redact(account.id, "a1");
    assert_json_snapshot!(redactor.response_json(&mut response).await.unwrap(), {
        ".data.attributes.created_at" => "[datetime]",
        ".data.attributes.updated_at" => "[datetime]",
    });
}
