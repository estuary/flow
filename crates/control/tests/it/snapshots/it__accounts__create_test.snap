---
source: crates/control/tests/it/accounts.rs
assertion_line: 38
expression: redactor.response_json(response).await.unwrap()

---
{
  "data": {
    "attributes": {
      "created_at": "[datetime]",
      "display_name": "Bruce Wayne",
      "email": "batman@batcave.test",
      "id": "[a1]",
      "name": "batman",
      "unique_name": "batman",
      "updated_at": "[datetime]"
    },
    "id": "[a1]",
    "links": {
      "self": "http://127.0.0.1:0/accounts/[a1]"
    },
    "type": "account"
  }
}
